sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/GraphqlSchemaFactory.java

```diff
@@ -0,0 +1,391 @@
+            LinkedHashMap::new, Collectors.toList()));
+  }
+
+  public GraphQLSchema generate() {
+    for (Map.Entry<NamePath, List<SqrlTableMacro>> path : fieldPathToTables.entrySet()) {
+      if (path.getKey().getLast().isHidden()) continue;
+
+      Optional<GraphQLObjectType> graphQLObjectType = generateObject(path.getValue(),
+          objectPathToTables.getOrDefault(path.getKey(), List.of()));
+      graphQLObjectType.map(objectTypes::add);
+    }
+
+    GraphQLObjectType queryType = createQueryType(objectPathToTables.get(NamePath.ROOT));
+
+    postProcess();
+
+    if (queryFields.isEmpty()) {
+      throw new RuntimeException("No tables found to build schema");
+    }
+
+    return GraphQLSchema.newSchema()
+        .query(queryType)
+        .additionalTypes(new LinkedHashSet<>(objectTypes))
+        .additionalType(CustomScalars.DATETIME)
+        .build();
+  }
+
+  private GraphQLObjectType createQueryType(List<SqrlTableMacro> relationships) {
+
+    List<GraphQLFieldDefinition> fields = new ArrayList<>();
+
+    for (SqrlTableMacro rel : relationships) {
+      String name = rel.getAbsolutePath().getDisplay();
+      if (name.startsWith(HIDDEN_PREFIX) || !isValidGraphQLName(name)) continue;
+      GraphQLFieldDefinition field = GraphQLFieldDefinition.newFieldDefinition()
+          .name(rel.getAbsolutePath().getDisplay())
+          .type(wrap(createTypeName(rel), rel.getMultiplicity()))
+          .arguments(createArguments(rel))
+          .build();
+      fields.add(field);
+    }
+
+    GraphQLObjectType objectType = GraphQLObjectType.newObject()
+        .name("Query")
+        .fields(fields)
+        .build();
+
+    usedNames.add("Query");
+
+    this.queryFields.addAll(fields);
+
+    return objectType;
+  }
+
+  private Optional<GraphQLObjectType> generateObject(List<SqrlTableMacro> tableMacros, List<SqrlTableMacro> relationships) {
+    //Todo check: The multiple table macros should all point to the same relnode type
+
+    Map<Name, List<SqrlTableMacro>> relByName = relationships.stream()
+        .collect(Collectors.groupingBy(g -> g.getFullPath().getLast(),
+            LinkedHashMap::new, Collectors.toList()));
+
+    SqrlTableMacro first = tableMacros.get(0);
+    RelDataType rowType = first.getRowType();
+    //todo: check that all table macros are compatible
+//    for (int i = 1; i < tableMacros.size(); i++) {
+//    }
+
+    List<GraphQLFieldDefinition> fields = new ArrayList<>();
+    for (RelDataTypeField field : rowType.getFieldList()) {
+      if (!relByName.containsKey(Name.system(field.getName()))) {
+        createRelationshipField(field).map(fields::add);
+      }
+    }
+
+    for (Map.Entry<Name, List<SqrlTableMacro>> rel : relByName.entrySet()) {
+      createRelationshipField(rel.getValue()).map(fields::add);
+    }
+
+    if (fields.isEmpty()) {
+      return Optional.empty();
+    }
+
+    String name = generateObjectName(first.getAbsolutePath());
```

Comment:
Instead of an object type, deduce the correct type w/ interfaces w/ is-a relationships